cpp/src/detail/permute_range.cu

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2020-2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cugraph/detail/collect_comm_wrapper.hpp>
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/detail/utility_wrappers.hpp>
+#include <cugraph/utilities/host_scalar_comm.hpp>
+#include <cugraph/utilities/shuffle_comm.cuh>
+
+#include <raft/core/device_span.hpp>
+#include <raft/core/handle.hpp>
+#include <raft/random/rng_state.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+
+namespace cugraph {
+
+namespace detail {
+
+template <typename vertex_t>
+rmm::device_uvector<vertex_t> permute_range(raft::handle_t const& handle,
+                                            raft::random::RngState& rng_state,
+                                            vertex_t local_range_size,
+                                            vertex_t local_range_start,
+                                            bool multi_gpu)
+{
+  if (multi_gpu) {
+    auto& comm           = handle.get_comms();
+    auto const comm_size = comm.get_size();
+    auto const comm_rank = comm.get_rank();
+
+    auto global_start =
+      cugraph::host_scalar_bcast(handle.get_comms(), local_range_start, 0, handle.get_stream());
+    auto sub_range_sizes =
+      cugraph::host_scalar_allgather(handle.get_comms(), local_range_size, handle.get_stream());
+    std::exclusive_scan(
+      sub_range_sizes.begin(), sub_range_sizes.end(), sub_range_sizes.begin(), global_start);
+    CUGRAPH_EXPECTS(
+      sub_range_sizes[comm_rank] == local_range_start,
+      "Invalid input arguments: a rage must have contiguous and non-overlapping values");
+  }
+  rmm::device_uvector<vertex_t> permuted_intergers(local_range_size, handle.get_stream());
+
+  // generate as many number as #local_vertices on each GPU
+  detail::sequence_fill(
+    handle.get_stream(), permuted_intergers.begin(), permuted_intergers.size(), local_range_start);
+
+  // shuffle/permute locally
```

Comment:
This step is not necessary for for multi gpu, the first thing you do in the multi-gpu case is to sort by a different random ordering.

I would suggest structuring this in the else block of the multi-gpu block.  So, if (MG) do the code currently in the MG, else do this block of code.

Reply:
Or you can separate the above multi GPU path to three blocks: 1) multi-GPU shuffling (if MG), 2) local shuffle (for both SG and MG), 3) taking care of extras/deficits (if MG). This will minimize code repetition.